wgpu/src/backend/mod.rs

```diff
@@ -10,14 +10,41 @@ mod web;
 pub(crate) use web::Context;
 
 #[cfg(any(
-    not(target_arch = "wasm32"),
+    all(
+        not(target_arch = "wasm32"),
+        any(
+            not(any(target_os = "macos", target_os = "ios")),
+            feature = "metal",
+            feature = "vulkan-portability",
+            feature = "angle"
+        )
+    ),
     target_os = "emscripten",
     feature = "webgl"
 ))]
 mod direct;
 #[cfg(any(
-    not(target_arch = "wasm32"),
+    all(
+        not(target_arch = "wasm32"),
+        any(
+            not(any(target_os = "macos", target_os = "ios")),
+            feature = "metal",
+            feature = "vulkan-portability",
+            feature = "angle"
+        )
+    ),
     target_os = "emscripten",
     feature = "webgl"
 ))]
 pub(crate) use direct::Context;
+
+#[cfg(all(
+    any(target_os = "macos", target_os = "ios"),
+    not(any(feature = "metal", feature = "vulkan-portability", feature = "angle"))
+))]
+mod dummy;
+#[cfg(all(
+    any(target_os = "macos", target_os = "ios"),
+    not(any(feature = "metal", feature = "vulkan-portability", feature = "angle"))
+))]
+pub(crate) use dummy::Context;
```

Comment:
I have made two separate commits for this so we can easily remove it if we want to.

Wgpu is used by a lot of libraries, so failing compilation when no backend is selected will make it harder for libraries to use `default-features = false`, which is what we should encourage, so end-users can select backends they want instead of libraries doing it for them.

Reply:
Alternatively to just panicking with `unimplemented!`, we could actually return an error that no device was found. But that seems worse to me, unless we clarify the error to also explain that there might be adapters available but no backend is enabled for them.

Reply:
not so picky about how a runtime placeholder backend would fail exactly. Just the fact that we have to duplicate the entire surface for a third time makes me unhappy.

> Wgpu is used by a lot of libraries, so failing compilation when no backend is selected will make it harder for libraries to use default-features = false, which is what we should encourage, so end-users can select backends they want instead of libraries doing it for them.

I mean this totally makes sense, but the question is whether there is any usecase at all for not having a valid backend and I doubt there is. Therefore, why should one wait for compilation to finish and running to find out that the build configuration was broken in the first place.

Reply:
I don't disagree that there is a tradeoff. This is a problem many libraries have faced and I don't know if any popular consensus was reached.

> I mean this totally makes sense, but the question is whether there is any usecase at all for not having a valid backend and I doubt there is. Therefore, why should one wait for compilation to finish and running to find out that the build configuration was broken in the first place.

The only advantage I see is to allow library developers to depend on Wgpu without having to select a backend.
See a recent discussion in https://github.com/rust-windowing/winit/issues/3174 and a comment by @cwfitzgerald https://github.com/rust-windowing/winit/issues/3174#issuecomment-1775626177.

I have to say that initially I was in favor of supporting compilation with `default-features = false`, but now that I think about it a bit more I'm unsure. It seems simple to me for library developers to enable `default` in `dev-dependencies` to be able to work with it. End-users have to select a backend anyway, so getting the compilation error is even helpful in this case.

Right now at least I'm back to removing this, let me know how you want me to proceed.

Reply:
Fixed in https://github.com/gfx-rs/wgpu/pull/4815/commits/8a6ca6de06867760c7bdf1368a1831b64586b703.